rfd/0183-access-request-kube-resource-allow-list.md

```diff
@@ -0,0 +1,247 @@
+
+Admins can specify in a role, a Kubernetes request mode option that requires users to select [Kubernetes subresources](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  options:
+    request_mode:
+      # Disables requesting whole Kubernetes cluster, but can request any of the kube subresources
+      kubernetes: ['resource']
+```
+
+User requesting kube_cluster will be denied.
+
+## As an admin, I want to limit Kubernetes resource requesting to pods and namespaces only
+
+Admins can specify in a role, a Kubernetes request mode option that requires users to select either pods or namespaces.
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  options:
+    request_mode:
+      # Can request only pods or namespaces
+      kubernetes: ['pod', 'namespace']
+```
+
+User requesting any other Kubernetes kind will get denied.
+
+# Details
+
+## New request field for role spec
+
+We will introduce a new role option under the `options` section named `request_mode`:
+
+```yaml
+kind: role
+metadata:
+  name: role-name
+spec:
+  options:
+    # new field
+    request_mode:
+      kubernetes: [<list of kinds>]
+```
+
+The `options.request_mode.kubernetes` list will accept the same const we use in our backend [kube_cluster](https://github.com/gravitational/teleport/blob/5924c97d7fc831aeed20990e4c72fe6a0bc95fba/api/types/constants.go#L203) and any of its [Kubernetes subresource kinds](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
+
+An admin may not care what kind of subresources a user requests, as long as the user is not requesting the whole Kubernetes cluster. For this case we will introduce a separate const `resource` that means a user can request to any Kubernetes subresources.
+
+| Request Mode Values                           |                           Explanation                            |
+| --------------------------------------------- | :--------------------------------------------------------------: |
+| []                                            |                         no restrictions                          |
+| `options.request_mode.kubernetes` not defined |                         no restrictions                          |
+| [kube_cluster, resource]                      |                         no restrictions                          |
+| [resource]                                    | requires users to request for any of the Kubernetes subresources |
+| [kube_cluster]                                |    requires users to request for the whole Kubernetes cluster    |
+| [namespace]                                   |          requires users to request for only namespaces           |
+| [namespace, pod]                              |      requires users to request for only namespaces OR pods       |
```

Comment:
The general idea makes sense but I feel like this is too many values tbh, I would scope this down to two possible options for now: `kube_cluster` and `namespace`. Then from user's perspective it will be clear what they can either allow someone to request access to cluster or namespace or both. We can always add more options later if needed but let's start simple.

Reply:
replaced with sasha's recommendation